Shrink released binaries using stripping and UPX compression

zcahana opened the pull request:
The following PR applies two methods to significantly shrink the size of our released binaries, based on the [following method](https://blog.filippo.io/shrink-your-go-binaries-with-this-one-weird-trick/).
1. Strip symbol tables and debug information using the `-s -w` [linker flags](https://golang.org/cmd/link/).
  This shrinks the binary by about %25-%50, while still keeping types/methods/filepaths to be displayed in panic stack traces.
2. Compress the binary into a self-decompressing executable using [UPX](https://upx.github.io/):
  This further shrinks the binary by an additional %80, to a total of about %90!
  The cost is the decompression latency when invoking the executable, which ranges between 100-400ms, depending on binary size and compression method.
  
Note that binary size was not much of a concern until now, being around 10-11 MB per binary.
But, with the reimplementation of the kubernetes adapter using the [official kubernetes go client](https://github.com/kubernetes/client-go) (in a separate PR), the `a8registry` & `a8sidecar` binaries sizes have skyrocketed to ~50 MB.
Combining the two methods above, the binary sizes are significantly reduced:

```bash
zvic@zvic-vbox:~/Workspace/go/src/github.com/amalgam8/amalgam8/bin (compress_binaries)$ ll -h
total 88M
drwxrwxr-x  2 zvic zvic 4.0K Nov 23 15:56 ./
drwxrwxr-x 20 zvic zvic 4.0K Nov 23 15:49 ../
-rwxrwxr-x  1 zvic zvic  46M Nov 23 11:27 a8sidecar*
-rwxrwxr-x  1 zvic zvic  26M Nov 23 11:24 a8sidecar-stripped*
-rwxrwxr-x  1 zvic zvic 6.0M Nov 23 11:24 a8sidecar-stripped-upx*
-rwxrwxr-x  1 zvic zvic 5.7M Nov 23 11:24 a8sidecar-stripped-upx-best*
-rwxrwxr-x  1 zvic zvic 4.5M Nov 23 11:24 a8sidecar-stripped-upx-best-ultra*
```

- `a8sidecar` (our vanilla sidecar) as well as `a8sidecar-stripped` starts up in about 50 ms.
- `a8sidecar-stripped-upx` (default compression method) as well as `a8sidecar-stripped-upx-best` (best compression method) starts up in about 150 ms.
- `a8sidecar-stripped-upx-best-ultra` (better than best compression :smile:) starts up in about 450 ms.

Two additional notes:
- According to the UPX docs, the produced executables are self-containing (no dependencies, statically linked) are should be able to run on any linux platform capable of running ELF executables (i.e., any).
- We need to make sure there are no issues with [UPX license](https://github.com/upx/upx/blob/master/LICENSE). For my understanding, the only section which applies to us is [SPECIAL EXCEPTION FOR COMPRESSED EXECUTABLES](https://github.com/upx/upx/blob/master/LICENSE#L68), which doesn't apply any special restrictions or requirements on us. @ericvn I'd appreciate if you could take a look and see if we can clear that.

zcahana committed:
Shrink released binaries using stripping and UPX compression

ijsnellf reviewed (approved):
LGTM, but I don't know about the UPX license. I'll assign @ericvn to review.

rshriram commented:
@zcahana any idea what the startup latency with uncompressed executable ?

zcahana commented:
@rshriram that's about 50-60ms (on my dev vm)

rshriram commented:
So we need to be careful about the size vs startup time trade offs. While
increased size results in longer time to download the image, once
downloaded, the image gets cached. So the download time is amortized over
time. Otoh, adding an extra 100ms startup latency seems to be a high
overhead, for each container in the system irrespective of caching. It
leaves no room for optimization.

At the end of the day, the uncompressed binary is the one that executes.
Given that these are code pages, they are most probably shared as well in
the kernels buffer cache (except for btrfs and devicemapper which are bad
choices anyway). So the startup overhead caused by reading a 100mb file (if
any) will likely go down as these code pages already exist.

If we can cut down something else in the docker image, great. If not, we
just live with the added bloat. It's a minor inconvenience, but certainly
no performance loss.
On Wed, Nov 23, 2016 at 3:15 PM Zvi Cahana notifications@github.com wrote:

> @rshriram https://github.com/rshriram that's about 50-60ms (on my dev
> vm)
> 
> —
> You are receiving this because you were mentioned.
> 
> Reply to this email directly, view it on GitHub
> https://github.com/amalgam8/amalgam8/pull/418#issuecomment-262619880,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AH0qdzqIEW5dWY7G8qGy7fslKhgmI9Hdks5rBJ7EgaJpZM4K6vyI
> .
> 
> ## 

~shriram

zcahana committed:
Merge branch 'master' into shrink_binaries

zcahana committed:
Removed --ultra-brute UPX flag

zcahana commented:
I agree that we need to consider the size vs startup time tradeoff, but disagree about some of the above statements:

- Image download is not a "one time" thing, not even when caches are involved. Images gets changed, pushed, pulled, moved across hosts, moved across dev/stage/prod environments all the time.
- 100ms additional overhead, IMO, is negligible. The typical deployment model for the sidecar/registry is with long-running servers and not short-lived jobs. Additionally, such servers would normally start "in the background" until they're ready to receive traffic. Just compare it to a Liberty (or, tWAS) server startup time (seconds? minutes?) :smile: 

I agree that 450ms is probably too much with respect to the size-startup time tradeoff.
I added another commit which removes the `--ultra-brute` compression flags. This increases the binary size from 4.5MB to 5.7MB (negligible too), but reduces the startup time from ~450ms to ~150ms.

P.S., running some more experiments, it seems that the average startup time of our non-compressed sidecar is ~70ms, so the overhead is only about 80ms.

zcahana commented:
Travis build is failing... I believe this is due to issue #401 which is still open.

zcahana committed:
Add linker flags to enable static linking

zcahana commented:
@ericvn did you get the chance to look at the license? If it's good, I think we can go ahead and merge this.

rshriram commented:
```
Images gets changed, pushed, pulled, moved across hosts, moved across dev/stage/prod environments all the time.
```
That may be true for app images/docker based environments. But not for the sidecar container running in a kubernetes environment. The sidecar image is relatively static.

Alternative approach would be to add documentation on how to reduce image sizes using UPX as an alternative. This would avoid licensing issues, while allowing image-size conscious users to customize it on their own, should they wish to, rather than us forcing it upon them. 

In K8S, I think the proponents of "reducing the pod startup latency" would strongly be against this 80ms overhead in exchange for compressed binary size (5Mb). Because ultimately, the uncompressed binary is the thing that will be executing and if its going to take up 50Mb, nothing can be done about it. 

The second thing is that this kills memory sharing. As I said earlier, even if I had 100 of these binaries on the same host, the effective memory used would be 50Mb plus heap memory for individual processes. With the UPX approach, we are looking at 100 * 50Mb plus heap memory for individual processes.

zcahana committed:
Removed UPX compression from Makefile release targets

zcahana commented:
Although I'm not in agreement about the latency overhead, the memory sharing thing is indeed an issue.
So, I've removed the UPX compression step from the Makefile release targets. The compression targets are still in place, but are not invoked by our automation. They can be invoked manually if anyone wishes to.

zcahana committed:
Merge branch 'master' into shrink_binaries

rshriram commented:
Great. LGTM

rshriram commented:
Just make sure to document this in the developer documentation at least (about reducing binary sizes).
